src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -866,12 +872,21 @@ internal unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase()
 
                 if (length > 0)
                 {
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(ptr[0]))
+                    {
+                        goto NotAscii;
+                    }
+
                     // Where length is 4n-3 (e.g. 1,5,9,13,17) this additionally consumes the null terminator
                     hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[0] | NormalizeToLowercase);
                 }
 
                 return (int)(hash1 + (hash2 * 1566083941));
             }
+
+            // Slow Non-ASCII fallback
+            NotAscii:
+            return Marvin.ComputeHash32OrdinalIgnoreCase(ref _firstChar, _stringLength, 0, 0); // zero seed to make it non-random
```

Comment:
Since we know this is going to fall into the non-ASCII code path, I'd recommend doing some targeted duplication of the logic from `Marvin.ComputeHash32OrdinalIgnoreCaseSlow`. Fall down a separate code path which either stackallocs a temp buffer or rents a new array, normalize the current string to uppercase inside that buffer, then run the naïve (non-Marvin) logic inside that buffer. It should keep most of the perf gains of this routine, even for non-ASCII data.

Reply:
Ah right, initially I just wanted to test it on CI in general but looks like it works so ok, will optimize the fallback a little like you suggest